### PR TITLE
3957 - Stop RUCSS Processing noscript content

### LINKED ISSUE #3957: RUCSS problem with <noscript> tags
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version
- Used the search feature to ensure that the bug hasn’t been reported before

**Describe the bug**
The style that's present in the <noscript> is processed by the Treeshaker (and removed from the source).

**To Reproduce**
1. Add the following code to the `head`:
`<noscript><style id="test">div{display:none !important;}</style></noscript>`
2. Enable RUCSS
3. Visit the page

**Expected behavior**
We should:
1. Always preserve the `<noscript><style id="id"></style></noscript>` in the source
2. Exclude it from processing, so nothing from this file is added to the used CSS

**Additional context**
Adding `high` priority, as popular social [sharing plugin](https://wordpress.org/plugins/social-pug/) has `data-lazy-src` in the JS file. It ends up in a conflict with the Lazy Load, as  `[data-lazy-src]{display:none !important;}` from the <noscript> is added to the Used CSS

**Backlog Grooming (for WP Media dev team use only)**
- [x] Reproduce the problem
- [x] Identify the root cause
- [x] Scope a solution
- [x] Estimate the effort

### PR DESCRIPTION
**Pull request by iCaspar**
## Description

Closes #3957 

## Type of change

- [X] Bug fix (non-breaking change which fixes an issue)

## Is the solution different from the one proposed during the grooming?

As Groomed

## How Has This Been Tested?

Testcase for noscript content added to `Subscriber\treeshake` fixture.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: inc/Engine/Optimization/RUCSS/Controller/UsedCSS.php [hunk 1]
```diff
@@ -418,9 +417,25 @@ private function update_used_css( int $id, array $data ) {
...
-		$link_styles        = $this->find( '<link\s+([^>]+[\s"\'])?href\s*=\s*[\'"]\s*?(?<url>[^\'"]+\.css(?:\?[^\'"]*)?)\s*?[\'"]([^>]+)?\/?>', $html_nocomments );
-		$inline_styles      = $this->find( '<style(?<atts>.*)>(?<content>.*)<\/style>', $html_nocomments );
+		$html_nocomments       = $this->hide_comments( $html );
+		$link_style_pattern    = '<link\s+([^>]+[\s"\'])?href\s*=\s*[\'"]\s*?(?<url>[^\'"]+\.css(?:\?[^\'"]*)?)\s*?[\'"]([^>]+)?\/?>';
+		$inline_inline_pattern = '<style(?<atts>.*)>(?<content>.*)<\/style>';
+
+		$link_styles   = $this->find(
+			'<noscript[^>]*>.*' .
+			'<link\s+([^>]+[\s"\'])?href\s*=\s*[\'"]\s*?([^\'"]+\.css(?:\?[^\'"]*)?)\s*?[\'"]([^>]+)?\/?>' .
+			'.*<\/noscript> (*SKIP)(*FAIL)|' .
```
Comment (crystinutzaa) [inline inc/Engine/Optimization/RUCSS/Controller/UsedCSS.php]: There is an extra space between noscript and SKIP `'.*<\/noscript> (*SKIP)(*FAIL)|' .` 
Will this cause any issues?

File: tests/Fixtures/inc/Engine/Optimization/RUCSS/Frontend/Subscriber/treeshake.php [hunk 2]
```diff
@@ -235,6 +235,62 @@
...
 
+		'shouldNotProcessItemsInsideNoscriptTag' => [
+			'config'       => [
+				'html'                  => '<!DOCTYPE html>
+<html>
+<head>
+	<meta charset="utf-8">
+	<title>My Awesome Page</title>
+	<style>h2{color:blue;}</style>
+	<noscript><style>h3{color:green;}</style></noscript>
```
Comment (crystinutzaa) [inline tests/Fixtures/inc/Engine/Optimization/RUCSS/Frontend/Subscriber/treeshake.php]: Can we add a testcase for links inside noscript tags too?

**Review by crystinutzaa — APPROVED** [review]

**Comment by iCaspar** [comment]
@Mai-Saad The SaaS end of this is merged onto treeshaker, so you should be able to QA this now using the SaaS develop server.

**Comment by Mai-Saad** [comment]
@iCaspar @Tabrisrp working fine for the script mentioned on ticket steps but not working probably with the following formats
`< noscript><style id="test">div{display:none !important color:red;}</style></noscript>` and `<noscript><style id="test">div{display:none !important color:red;}</style>< / noscript >` => Donn we need to handle these formats? (not sure if they are correct formats)

**Comment by GeekPress** [comment]
@Mai-Saad `< noscript>` this isn't a valid tag format. If something is used like that it will be printed on the website as the browser won't consider this as a valid tag.

Here an example with a wrong formatted `<p>` tag:
https://share.getcloudapp.com/2Nuqwbyv

As you can notice `< p>` is printed as it's not a valid tag for the browser.

Same behavior for the closure:
https://share.getcloudapp.com/04u2A6NK

**Comment by remyperona** [comment]
Indeed, base on the HTML specifications, you can't have space before the tag name, both in opening and closing tags.

The opening `<` character must always be followed by the tag name directly.

You can have any number of space before the closing `>` character though.

**Review by Mai-Saad — APPROVED** [review]
Working as expected 
[testrail-report-196.pdf](https://github.com/wp-media/wp-rocket/files/6677815/testrail-report-196.pdf)
Notes: as agreed not related to this issue https://github.com/wp-media/wp-rocket/issues/3957#issuecomment-857697246

**Pull request merged**